File: android/src/main/java/id/animo/secure/environment/ExpoSecureEnvironmentModule.kt
package id.animo.secure.environment

import android.os.Build
import androidx.annotation.RequiresApi
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class ExpoSecureEnvironmentModule : Module() {
    @RequiresApi(Build.VERSION_CODES.R)
    override fun definition() =
        ModuleDefinition {
            Name("ExpoSecureEnvironment")

            Function("generateKeypair") { id: String, biometricsBacked: Boolean ->
                SecureEnvironment.generateKeypair(appContext, id, biometricsBacked)
            }

            Function("getPublicBytesForKeyId") { keyId: String ->
                return@Function SecureEnvironment.getPublicBytesForKeyId(appContext, keyId)
            }

            Function("supportsSecureEnvironment") {
                return@Function SecureEnvironment.supportsSecureEnvironment(appContext)
            }

            AsyncFunction("sign") { id: String, message: ByteArray, biometricsBacked: Boolean, promise: Promise ->
                SecureEnvironment.sign(appContext, id, message, biometricsBacked, promise)
            }

            Function("deleteKey") { id: String ->
                return@Function SecureEnvironment.deleteKey(appContext, id)
            }
        }
}

File: ios/ExpoSecureEnvironmentModule.swift
import ExpoModulesCore

public class ExpoSecureEnvironmentModule: Module {
  public func definition() -> ModuleDefinition {
    Name("ExpoSecureEnvironment")

    Function("generateKeypair") { (id: String, biometricsBacked: Bool) -> Void in
      try SecureEnvironment.generateKeyPair(id, biometricsBacked)
    }

    // Expo minimal version always supports the secure environment for iOS
    Function("supportsSecureEnvironment") { () -> Bool in
      return true
    }

    Function("getPublicBytesForKeyId") { (keyId: String) -> Data in
      return try SecureEnvironment.getPublicBytesForKeyId(keyId)
    }

    Function("deleteKey") { (keyId: String) in
      return try SecureEnvironment.deleteKey(keyId)
    }

    AsyncFunction("sign") {
      (id: String, message: Data) async throws -> Data in
      return try await SecureEnvironment.sign(id, message)
    }
  }
}
